pkg/gatewayserver/io/basicstationlns/messages/upstream.go

```diff
@@ -0,0 +1,294 @@
+func (updf *UplinkDataFrame) ToUplinkMessage(ids ttnpb.GatewayIdentifiers, bandID string) (ttnpb.UplinkMessage, error) {
+	up := ttnpb.UplinkMessage{}
+	up.ReceivedAt = time.Now()
+
+	parsedMHDR := ttnpb.MHDR{}
+	err := lorawan.UnmarshalMHDR([]byte{byte(updf.MHdr)}, &parsedMHDR)
+	if err != nil {
+		return ttnpb.UplinkMessage{}, errUplinkDataFrame.WithCause(err)
+	}
+	if (parsedMHDR.MType != ttnpb.MType_UNCONFIRMED_UP) && (parsedMHDR.MType != ttnpb.MType_CONFIRMED_UP) {
+		return ttnpb.UplinkMessage{}, errUplinkDataFrame
+	}
+
+	micBytes, err := getInt32AsByteSlice(updf.MIC)
+	if err != nil {
+		return ttnpb.UplinkMessage{}, errUplinkDataFrame.WithCause(err)
+	}
+
+	var fport uint32
+	if updf.FPort == -1 {
+		fport = 0
+	} else {
+		fport = uint32(updf.FPort)
+	}
+
+	devAddrBytes, err := getInt32AsByteSlice(updf.DevAddr)
+	if err != nil {
+		return ttnpb.UplinkMessage{}, errUplinkDataFrame.WithCause(err)
+	}
+	devAddr := types.DevAddr{}
+	err = devAddr.UnmarshalBinary(devAddrBytes)
+	if err != nil {
+		return ttnpb.UplinkMessage{}, errUplinkDataFrame.WithCause(err)
+	}
+
+	fctrl := ttnpb.FCtrl{}
+	err = lorawan.UnmarshalFCtrl([]byte{byte(updf.FCtrl)}, &fctrl, true)
+	if err != nil {
+		return ttnpb.UplinkMessage{}, errUplinkDataFrame.WithCause(err)
+	}
+
+	up.Payload = &ttnpb.Message{
+		MHDR: parsedMHDR,
+		MIC:  micBytes,
+		Payload: &ttnpb.Message_MACPayload{MACPayload: &ttnpb.MACPayload{
+			FPort:      fport,
+			FRMPayload: []byte(updf.FRMPayload),
+			FHDR: ttnpb.FHDR{
+				DevAddr: devAddr,
+				FCtrl:   fctrl,
+				FCnt:    uint32(updf.FCnt),
+				FOpts:   []byte(updf.FOpts),
+			},
+		}},
+	}
+
+	up.RawPayload, err = lorawan.MarshalMessage(*up.Payload)
+	if err != nil {
+		return ttnpb.UplinkMessage{}, errUplinkDataFrame.WithCause(err)
+	}
+
+	timestamp := uint32(updf.RadioMetaData.UpInfo.XTime & 0xFFFFFFFF)
+
+	ulToken := ttnpb.UplinkToken{
+		GatewayAntennaIdentifiers: ttnpb.GatewayAntennaIdentifiers{
+			GatewayIdentifiers: ids,
+			AntennaIndex:       uint32(updf.RadioMetaData.UpInfo.RCtx),
+		},
+		Timestamp: timestamp,
+	}
+	ulTokenBytes, err := ulToken.Marshal()
+	if err != nil {
+		return ttnpb.UplinkMessage{}, errJoinRequestMessage.WithCause(err)
+	}
+
+	rxTime := time.Unix(updf.RadioMetaData.UpInfo.RxTime, 0)
+	rxMetadata := &ttnpb.RxMetadata{
+		GatewayIdentifiers: ids,
+		Time:               &rxTime,
+		Timestamp:          timestamp,
+		RSSI:               updf.RadioMetaData.UpInfo.RSSI,
+		SNR:                updf.RadioMetaData.UpInfo.SNR,
+		UplinkToken:        ulTokenBytes,
+	}
+	up.RxMetadata = append(up.RxMetadata, rxMetadata)
+
+	loraDR, err := getDataRateFromIndex(bandID, updf.RadioMetaData.DataRate)
+	if err != nil {
+		return ttnpb.UplinkMessage{}, errUplinkDataFrame.WithCause(err)
+	}
+	up.Settings = ttnpb.TxSettings{
+		Frequency:     updf.RadioMetaData.Frequency,
+		DataRateIndex: ttnpb.DataRateIndex(updf.RadioMetaData.DataRate),
```

Comment:
GS doesn't officially know this value at this point, see UDP frontend frontends. This gets filled out by NS upstream